### PR TITLE
Expression framework refactoring

### PR DESCRIPTION
**Pull request by ben-clayton**
In preparation for adding new Array type and value.

Replace the single `Scalar` value with a separate class per value type. This removes a bunch of runtime type dispatching for separate classes.
Move the `TypeXXX` constructed types to a new `Type` interface (namespace).
Add the new `ArrayType` and `ArrayValue` types.

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [x] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/call/builtin/distance.spec.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 export const description = `
 Execution tests for the 'distance' builtin function
 
-S is AbstractFloat, f32, f16
```
Comment (zoddicus) [inline src/webgpu/shader/execution/expression/call/builtin/distance.spec.ts:4]: Looks like whatever search-replace was used was slightly overmatching

File: src/webgpu/util/floating_point.ts [hunk 2]
```diff
@@ -1074,7 +1074,7 @@ export abstract class FPTraits {
   /** @returns 1 * ULP: (number) */
   public abstract readonly oneULP: (target: number, mode?: FlushMode) => number;
   /** @returns a builder for converting numbers to Scalars */
```
Comment (zoddicus) [inline src/webgpu/util/floating_point.ts:1076]: nit: ScalarValues

File: src/webgpu/util/conversion.ts [hunk 3]
```diff
@@ -612,8 +616,12 @@ export class ScalarType {
     return this._size;
   }
 
+  public get alignment(): number {
+    return this._size;
+  }
+
   /** Constructs a Scalar of this type with `value` */
```
Comment (zoddicus) [inline src/webgpu/util/conversion.ts:623]: nit: ScalarValue

**Review by zoddicus — COMMENTED** [review]
overall the change looks good.

Some comment/document issues. In the descriptions from the spec AbstractInt and AbstractFloat have been replaced with Type.XXX form, which I don't think is correct. Also there is some places where the type name Scalar in comments didn't get changed to ScalarValue.

I just identified a few locations of these problems, and didn't systematically go through the entire change, since there is probably a way to automate fixing this.

**Review by zoddicus — APPROVED** [review]
Aprroving, assuming the changes gets cleaned up

File: src/webgpu/shader/execution/expression/call/builtin/distance.spec.ts (hunk 1 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/call/builtin/distance.spec.ts:4]: Done

File: src/webgpu/util/floating_point.ts (hunk 2 above)
Comment (ben-clayton) [reply src/webgpu/util/floating_point.ts]: Done

File: src/webgpu/util/conversion.ts (hunk 3 above)
Comment (ben-clayton) [reply src/webgpu/util/conversion.ts]: Done

**Pull request merged**